### ISSUE TITLE
Generation not working properly

### ISSUE THREAD

**Issue by TheMrguiller** [lvwerra/trl#333]
Hi,

I am following the exact same example provided in sentiment.py but adapted to my subject, toxicity. When the model is generating the responses i get random words that makes no sense. Something like the following:
```
finfinbalfinfinfinborfinbalborfinboranfinfinagfinantolerfinborgerfinborbalborborbalfinbalbalfinnicfinanborborfinfinanfinrefined finrefined borfinetborfinacfinbalnicbalborananborbaltolerborbornicbalbalbalagfinattfinacborborboracrefined anfinkinborrefined boranborfinananbalfinbormeter borkinananmeter borfinrefined anan__end__
```
But when i use the Hugginface code for generation on my Blenderbot model, i get the following:
```
input_ids = tokenizer.encode('I enjoy walking with my cute dog', return_tensors='pt')
generate text until the output length (which includes the context length)
greedy_output = model.generate(input_ids, max_length=50)
print("Output:\n" + 100 * '-')
print(tokenizer.decode(greedy_output[0], skip_special_tokens=True))
Output:what kind of dog do you have? i don't own a dog, but i'd love to have one someday.
```
In order to train it i used the following class:
```
from dataclasses import dataclass, field
from typing import Optional

import torch
from datasets import load_dataset
from tqdm import tqdm
from transformers import AutoTokenizer, HfArgumentParser, pipeline
from datasets import Dataset
import sys
sys.path.append("/home/d4k/Documents/guillermo/trl")
from trl.models import AutoModelForCausalLMWithValueHead
from trl.core import LengthSampler, set_seed
from trl.trainer import PPOConfig, PPOTrainer
import pandas as pd
from reward import ToxicityRewardFunction
tqdm.pandas()
#Primero configuramos el PPOConfig
config = PPOConfig(
    model_name="TheMrguiller/Blenderbot_Prosocial",
    learning_rate=1e-6,
    log_with="wandb",
    batch_size=64,
    tracker_project_name="trl",
    mini_batch_size=16,
)
sent_kwargs = {"return_all_scores": True, "function_to_apply": "none", "batch_size": 64}
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
# Below is an example function to build the dataset. In our case, we use the IMDB dataset
# from the `datasets` library. One should customize this function to train the model on
# its own dataset.
def build_dataset(input_min_text_length=5, input_max_text_length=512,split="train"):
    """
    Build dataset for training. This builds the dataset from `load_dataset`, one should
    customize this function to train the model on its own dataset.

    Args:
        dataset_name (`str`):
            The name of the dataset to be loaded.

    Returns:
        dataloader (`torch.utils.data.DataLoader`):
            The dataloader for the dataset.
    """
    tokenizer = AutoTokenizer.from_pretrained("facebook/blenderbot_small-90M")
    tokenizer.pad_token = tokenizer.eos_token
    # load imdb with datasets
    if split=="train":
        df = pd.read_csv("/home/d4k/Documents/guillermo/trl/examples/toxicity/data/train_blenderbot_RL.csv")
    else:
        df = pd.read_csv("/home/d4k/Documents/guillermo/trl/examples/toxicity/data/test_blenderbot_RL.csv")
    df = df.sample(frac=1, random_state=456)
    ds = Dataset.from_pandas(df)
    ds = ds.rename_columns({"com": "entrada"})
    

    input_size = LengthSampler(input_min_text_length, input_max_text_length)

    def tokenize(sample):
        sample["input_ids"] = tokenizer.encode(sample["entrada"],truncation=True)
        sample["query"] = tokenizer.decode(sample["input_ids"])
        return sample

    ds = ds.map(tokenize, batched=False)
    ds.set_format(type="torch")
    return ds

dataset = build_dataset(input_min_text_length=5,input_max_text_length=512)

def collator(data):
    return dict((key, [d[key] for d in data]) for key in data[0])

set_seed(42)

# Now let's build the model, the reference model, and the tokenizer.
model = AutoModelForCausalLMWithValueHead.from_pretrained(config.model_name).to(device)
ref_model = AutoModelForCausalLMWithValueHead.from_pretrained(config.model_name).to(device)
tokenizer = AutoTokenizer.from_pretrained("facebook/blenderbot_small-90M")
tokenizer.pad_token = tokenizer.eos_token

ppo_trainer = PPOTrainer(config, model, ref_model, tokenizer, dataset=dataset, data_collator=collator)

toxicity=ToxicityRewardFunction()

# We then define the arguments to pass to the `generate` function. These arguments
# are passed to the `generate` function of the PPOTrainer, which is a wrapper around
# the `generate` function of the trained model.
generation_kwargs = {
    "min_length": -1,
    "top_k": 0.0,
    "top_p": 1.0,
    "do_sample": True,
    "pad_token_id": tokenizer.eos_token_id,
}
output_min_length = 20
output_max_length = 128

output_length_sampler = LengthSampler(output_min_length, output_max_length)
print("Entrando en entrenamiento")
for epoch, batch in tqdm(enumerate(ppo_trainer.dataloader)):
    query_tensors = batch["input_ids"]

    # Get response from gpt2
    response_tensors = ppo_trainer.generate(
        query_tensors, **generation_kwargs
    )

    batch["response"] = tokenizer.batch_decode(response_tensors)
    print(batch["response"])
    # Compute sentiment score
    texts_context = ["[HST]"+q+"[ANS]"+ r for q, r in zip(batch["query"], batch["response"])]
    pipe_outputs=toxicity.calculate(batch["response"],texts_context )
    
    rewards = [torch.tensor(output) for output in pipe_outputs]
    # print(f"Rewards:{rewards}")
    # Run PPO step
    stats = ppo_trainer.step(query_tensors, response_tensors, rewards)
    ppo_trainer.log_stats(stats, batch, rewards)
```
And to launch it, i am using the following:
```
python blenderbot-toxicity.py
```
@younesbelkada and @lvwerra  please give me a hand, thanks.

**Comment by TheMrguiller** [comment]
I fix the generation changing the class to AutoModelForSeq2SeqLMWithValueHead, but i am not sure if the collator is the proper one.

**Comment by younesbelkada** [comment]
Indeed you need to switch to `AutoModelForSeq2SeqLMWithValueHead` as blenderbot is an encoder-decoder model and not a decoder-only model.
Regarding the KL divergence issue, some users reported that batched generation for seq2seq models leads to negative KL divergence: https://github.com/lvwerra/trl/issues/256#issuecomment-1503006078 
So I would maybe first try without batched generation and see if you observe the same behavior

**Comment by TheMrguiller** [comment]
Is there any other solution to the problem? Without batched generation the system will be slower in the training process

**Comment by avacaondata** [comment]
I'm experiencing similar issues @TheMrguiller and honestly I haven't found the cause for the negative KL divergence yet...

**Comment by TheMrguiller** [comment]
I tried debugging the code and I found some inconsistencies with the compute rewards. So I change the framework, until these elements are fixed.

**Comment by TheMrguiller** [comment]
No, its related to how they compute the KL Divergence and how is used in the compute reward. I am using RL4LMs but as i am not using PEFT I didnt check if they support it.

**Comment by avacaondata** [comment]
Yeah I've used RL4LMs and it is very consistent, the negative KL divergence issue does not take place. However it's not as usable as trl, at least from my experience with it. I think it would be really useful if we are able to solve this issue in trl, as it enables training with PEFT + Accelerate (making rl possible in consumer hardware) and is similar to transformers API, which makes the library easy to use. If you could please elaborate a little bit on the issue you found I would love to take a look at it and see if I can help in some way. Also pinning @lvwerra , maybe he can have an idea of how to solve it.

**Comment by TheMrguiller** [comment]
What do you mean by "not as usable as trl". I saw that when computing the reward, the reward that comes from the reward model is added in the last part of the tensor but is not added as a whole. When by the function definitions it should be R´=r -kl_coef*KL. It seems that the compute is not done the way it should. I am not an expert, but it seems strange to me.

**Comment by avacaondata** [comment]
I'm taking a look at what you said, specially these lines:  https://github.com/lvwerra/trl/blob/c85cdbdbd0d956bbc6f5a8e04b0036648480e7a1/trl/trainer/ppo_trainer.py#L936
As I understand it, in that function (compute_rewards) first kl penalty is computed: `-self.kl_ctl.value * kl` and then the score (pure reward) is added to the function: `reward[last_non_masked_index] += score`, which would be `R=-kl_coef*KL + r = r -kl_coef*KL`. Maybe I'm getting something wrong... Let's see if @lvwerra can help us clarify where this negative KL divergence issue might come from.

**Comment by lvwerra** [comment]
Yes, that's the correct way to compute the reward: the reward is only added to the last token in the sequence (similar to how an RL environment might only give the reward to the last step when e.g. a level is completed). In the `loss` function the reward is then discounted for all earlier steps. We haven't seen any issues on that side.

The reason for negative KL-div is usually related to a bug in generation. Can you check if the responses from the model before training make sense?

```python
response_tensors = ppo_trainer.generate(query_tensors, **generation_kwargs)
batch["response"] = tokenizer.batch_decode(response_tensors)
```

It could be that the way we generate batches with seq2seq models doesn't work properly for your model architecture.
